openedx/core/djangoapps/content_libraries/docs/decisions/0002-library-content-schema.rst

```diff
@@ -0,0 +1,115 @@
+      ],
+      "library_blocks": [
+        "lb:ScienceX:ChemLib1:problem:mc1",
+        "lb:EngineeringX:BasicChemLib:video:stiochiometry",
+        "lb:ScienceX:ChemLib1:problem:mc2",
+      ],
+      "field_overrides": [
+        "weight": 0.5
+      ],
+      "block_field_overrides": {
+        "lb:EngineeringX:BasicChemLib:video:stoichiometry": {
+          "display_name": "How to Stoichiometry"
+        }
+      }
+    }
+  }
+
+Accessing library-sourced blocks
+********************************
+
+Although the ``library_content_v2`` block described above stores references to libraries and library block usages, the content of the library-sourced blocks themselves would not be imported into in modulestore. Instead, their definitions would remain authoritatively within blockstore bundles. When rendering Studio-facing XBlock views, Studio would query blockstore for the requisite block OLX, deserialize it, and instantiate the block.
+
+We will also build an LMS-specific store for library blocks. When a blockstore content library is published, updated versions of the library blocks will be pushed into LMS. This store will be optimized specifically for the high-read LMS use case. It will be persisted so that if Studio/blockstore is unavailable, content can still be accessed by LMS.
+
+With this approach, we avoid having to copy blockstore library content to modulestore.
+
+
+Contextual block usage keys
+***************************
+
+The decision not to store library-sourced blocks in modulestore begs the question of what keys will be used to reference them.
+
+Content within the course run hierarchy is traditionally assigned a usage key of the form ``course-v1:ORG+COURSE+RUN+type@TYPE+block@USAGE_ID``. Since these library-sourced blocks are not stored along with the course run hierarchy in the normal way, though, I am not sure if this is appropriate.
+
+However, it is essential that library-sourced blocks usages are unambigously addressable. We cannot assume that a single library block will only be used in a course once; therefore, the library block usage keys of the form ``lb:ORG:LIBRARY:TYPE:ID`` are also not viable.
+
+Instead, we will introduce a new key namespace ``cb:``, roughly standing for either "course run block" or "contextual block". It will uniquely identify a usage of a defined block within a learning context. (The namespace was chosen to be analogous with ``lb:``, which identifies block definitions within blockstore content libraries.)
+
+A contextual block usage key will take the form::
+
+  cb::USAGE_CONTEXT::BLOCK_ID
+
+For example, if the stoichiometry video above were included in a ScienceX course run, its
+usage key would be look like::
+
+  cb::block-v1:ScienceX+Chem+101+2024+type@library_content+block@abcd1234::lb:EngineeringX:BasicChemLib:video:stoichiometry
+
+  clb:block-v1:ScienceX+Chem+101+2024+type@library_content+block@abcd1234:EngineeringX:BasicChemLib:video:stoichiometry
```

Comment:
To make these keys shorter, you could consider not encoding the full `lb:` key.
To find some of this content, the LMS would likely need to look up _both_ the `library_content` block (in modulestore) and the referenced library block (in blockstore). So since it's loading the `library_content` block anyways, it could load some map of shorter IDs to full library block keys.

```
clb:block-v1:ScienceX+Chem+101+2024+type@library_content+block@abcd1234:block01
clb:block-v1:ScienceX+Chem+101+2024+type@library_content+block@abcd1234:block02
... etc.
```